### PR TITLE
V2.0

### PR DESCRIPTION
**Pull request by igorshapiro**

### CODE REVIEW

File: lib/boot.js [hunk 1]
```diff
@@ -1,11 +1,13 @@
-var Hub = require('./hub')
+var config = require('./../config/config')
+var Hub = require('./hub/hub')
 var co = require('co')
 
 co(function*() {
-  var hub = new Hub()
+  console.log("Starting ServiceHub")
+  var hub = new Hub({ config: config })
   yield hub.start()
```
Comment (mikeys) [inline lib/boot.js]: @igorshapiro Is `start` actually defined on the `Hub` class? I couldn't find it...

Comment (igorshapiro) [inline lib/boot.js]: Me too :) Debugging why it doesn't fail

On Sat, Feb 20, 2016 at 4:03 PM, Miki Shiran <notifications@github.com>
wrote:

> In lib/boot.js
> <https://github.com/igorshapiro/hubjs/pull/18#discussion_r53559511>:
>
> >  var co = require('co')
> >
> >  co(function*() {
> > -  var hub = new Hub()
> > +  console.log("Starting ServiceHub")
> > +  var hub = new Hub({ config: config })
> >    yield hub.start()
>
> @igorshapiro <https://github.com/igorshapiro> Is start actually defined
> on the Hub class? I couldn't find it...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/igorshapiro/hubjs/pull/18/files#r53559511>.
>



-- 
Best Regards
Igor Shapiro
shapigor@gmail.com

File: lib/middlewares/recurring.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    var record = {
+      _id: msg.messageType,     // prevent registering multiple messages of the same type
+      recurringMessageId: shortid.generate(),
+      message: msg,
+      registeredAt: now,
+      nextAt: now,
+      prevAt: null,
+    }
+    var existingMsg = yield this.collection.findOneAsync({_id: msg.messageType})
+    if (existingMsg) {
```
Comment (tomeresk) [inline lib/middlewares/recurring.js]: Some type of upsert (maybe with findOneAndModify?) might work here

File: lib/middlewares/middleware.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    this.log = this.hub.log
+
+    this.service = options.service
+    this.params = options.params || {}
+    this.needs = {}
+    this.supports = {}
+  }
+
+  defineDependency(array, name, injectAs, service) {
+    if (injectAs.indexOf(':service') != -1) {
```
Comment (tomeresk) [inline lib/middlewares/middleware.js]: can use `.find` instead here

File: lib/middlewares/middleware.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  defineMandatoryDependency(name, injectAs, service) {
+    this.defineDependency(this.needs, name, injectAs, service)
+  }
+
+  defineOptionalDependency(name, injectAs, service) {
+    this.defineDependency(this.supports, name, injectAs, service)
+  }
+
+  getDependency(name, service) {
+    if (name.indexOf(':service') !== -1) {
```
Comment (tomeresk) [inline lib/middlewares/middleware.js]: can use `.find` instead here

File: lib/middlewares/recurring.js (hunk 2 above)
Comment (igorshapiro) [inline lib/middlewares/recurring.js]: Can't use it as it updates the whole document, while I need to update only the message part if it already exists (can't touch nextAt and prevAt)

File: lib/middlewares/middleware.js (hunk 3 above)
Comment (igorshapiro) [inline lib/middlewares/middleware.js]: `.includes`

File: lib/middlewares/recurring.js (hunk 2 above)
Comment (tomeresk) [inline lib/middlewares/recurring.js]: You can update only a part of the document with $set.
Change the update doc from `{ message: msg }` to `{ $set: { message: msg } }`

**Comment by mikeys** [comment]
@tomeresk  :+1: :+1: :+1:

**Pull request merged**